src/katgpucbf/xbgpu/kernels/beamform.mako

```diff
@@ -88,64 +94,78 @@ KERNEL REQD_WORK_GROUP_SIZE(BLOCK_SPECTRA, 1, 1) void beamform(
+        cplx accum[BATCH_BEAMS];
+        for (int i = 0; i < BATCH_BEAMS; i++)
+            accum[i] = make_float2(0.0f, 0.0f);
+
+        for (int a_batch = 0; a_batch < n_ants; a_batch += BATCH_ANTENNAS)
         {
-            for (int a = 0; a < batch_size; a++)
+            int a_batch_size = min(BATCH_ANTENNAS, n_ants - a_batch);
+            /* Precompute the weights for this batch.
+             * The work items collectively iterate over the interval
+             * [0, b_batch_size * a_batch_size).
+             */
+            for (int i = lid; i < b_batch_size * a_batch_size; i += BLOCK_SPECTRA)
             {
-                char4 sample = *in;
-                // Convert to float, and split the polarisations into an
-                // array.
-                cplx sample_pols[2] = {make_float2(sample.x, sample.y), make_float2(sample.z, sample.w)};
-                in += in_antenna_stride;
+                int beam = b_batch + i % b_batch_size;
+                int ant = a_batch + i / b_batch_size;
+                // Address into the global memory
+                int addr = ant * N_BEAMS + beam;
+                cplx w = weights[addr];
+                float cd = channel * delays[addr];
+                cplx rot;
+                sincospif(cd, &rot.y, &rot.x);
+                l_weights[i] = cmul(w, rot);
+            }
+            BARRIER(); // Complete all weight calculations before using them
 
-                /* It's critical that this loop gets unrolled, so that `pol` is
-                 * known at compile time.
-                 */
-#pragma unroll
-                for (int i = 0; i < N_BEAMS; i++)
+            if (valid)
+            {
+                for (int a = 0; a < a_batch_size; a++)
                 {
-                    int pol = beam_pols[i];
-                    accum[i] = cmad(l_weights[a * N_BEAMS + i], sample_pols[pol], accum[i]);
+                    char4 sample = *batch_in;
+                    // Convert to float, and split the polarisations into an
+                    // array.
+                    cplx sample_pols[2] = {make_float2(sample.x, sample.y), make_float2(sample.z, sample.w)};
+                    batch_in += in_antenna_stride;
+
+                    /* It's critical that this loop gets unrolled, so that `pol` is
+                     * known at compile time.
+                     */
+#pragma unroll
+                    for (int i = 0; i < b_batch_size; i++)
+                    {
+                        int beam = b_batch + i;
+                        int pol = beam_pols[beam];
+                        accum[i] = cmad(l_weights[a * b_batch_size + i], sample_pols[pol], accum[i]);
+                    }
                 }
             }
+            BARRIER();  // protect against next loop iteration overwriting l_weights
         }
-        BARRIER();  // protect against next loop iteration overwriting l_weights
-    }
-
-    if (!valid)
-        return;
```

Comment:
The removal of this branch of code makes me wonder what it was used for in the first place - seeing as it was checked so late in the day/kernel?